File: exercise/ts/day19/src/travel/santaTravelCalculator.ts
export class OverflowException extends Error {
    constructor(message: string) {
        super(message);
        this.name = "CustomOverflowException";
    }
}

export class SantaTravelCalculator {
    public static calculateTotalDistanceRecursively(numberOfReindeers: number): number {
        if(numberOfReindeers >= 32)
            throw new OverflowException(`Overflow for ${numberOfReindeers} reindeers`);

        if (numberOfReindeers === 1) return 1;

        return 2 * this.calculateTotalDistanceRecursively(numberOfReindeers - 1) + 1;
    }
}

File: exercise/python/day19/travel/SantaTravelCalculator.py
class SantaTravelCalculator:
    @staticmethod
    def calculate_total_distance_recursively(number_of_reindeers: int) -> int:
        if number_of_reindeers == 1:
            return 1
        if number_of_reindeers == 32:
            raise ArithmeticError("Integer overflow")

        result = 2 * SantaTravelCalculator.calculate_total_distance_recursively(number_of_reindeers - 1) + 1
        return result
